Platform/OpenNtfsDxe/Driver.h

```diff
@@ -0,0 +1,672 @@
+  UINT8  Padding[3];
+} INDEX_ROOT;
+
+typedef struct {
+  INDEX_ROOT Root;
+  UINT32     FirstEntryOffset;
+  UINT32     EntriesTotalSize;
+  UINT32     EntriesAllocatedSize;
+  UINT8      Flags;
+  UINT8      Padding[3];
+} ATTR_INDEX_ROOT;
+
+/**
+   Table 2.5. Layout of the $FILE_NAME (0x30) attribute
+   ____________________________________________________________________
+   Offset | Size | Description
+   --------------------------------------------------------------------
+   ~      | ~    | Standard Attribute Header
+   0x00   | 8    | File reference to the base record of the parent directory.
+   0x08   | 8    | C Time - File Creation
+   0x10   | 8    | A Time - File Altered
+   0x18   | 8    | M Time - MFT Changed
+   0x20   | 8    | R Time - File Read
+   0x28   | 8    | Allocated size of the file (a multiple of the Cluster size)
+   0x30   | 8    | Real size of the file (size of the unnamed $Data Attribute)
+   0x38   | 4    | Flags, e.g. Directory, compressed, hidden
+   0x3c   | 4    | Used by EAs and Reparse
+   0x40   | 1    | Filename length in characters (L)
+   0x41   | 1    | Filename namespace
+   0x42   | 2L   | File name in Unicode (not null terminated)
+   --------------------------------------------------------------------
+**/
+typedef struct {
+  UINT64 ParentDir;
+  UINT64 CreationTime;
+  UINT64 AlteredTime;
+  UINT64 ChangedMftTime;
+  UINT64 ReadTime;
+  UINT64 AllocatedSize;
+  UINT64 RealSize;
+  UINT32 Flags;
+  UINT32 Reparse;
+  UINT8  FilenameLen;
+  UINT8  Namespace;
+} ATTR_FILE_NAME;
+
+typedef struct {
+  INT32             Flags;
+  UINT8             *ExtensionMftRecord;
+  UINT8             *NonResAttrList;
+  UINT8             *Current;
+  UINT8             *Next;
+  UINT8             *Last;
+  struct _NTFS_FILE *BaseMftRecord;
+} NTFS_ATTR;
+
+typedef struct _NTFS_FILE {
+  UINT8                  *FileRecord;
+  UINT64                 DataAttributeSize;
+  UINT64                 AlteredTime;
+  UINT64                 Inode;
+  BOOLEAN                InodeRead;
+  NTFS_ATTR              Attr;
+  struct _EFI_NTFS_FILE  *File;
+} NTFS_FILE;
+
+typedef struct _EFI_NTFS_FILE {
+  EFI_FILE_PROTOCOL      EfiFile;
+  BOOLEAN                IsDir;
+  INT64                  DirIndex;
+  INT32                  Mtime;
+  CHAR16                 *Path;
+  CHAR16                 *BaseName;
+  UINT64                 Offset;
+  UINT32                 RefCount;
+  NTFS_FILE              RootFile;
+  NTFS_FILE              MftFile;
+  struct _EFI_FS         *FileSystem;
+} EFI_NTFS_FILE;
+
+typedef struct _EFI_FS {
+  EFI_SIMPLE_FILE_SYSTEM_PROTOCOL FileIoInterface;
+  EFI_FILE_PROTOCOL               EfiFile;
+  EFI_BLOCK_IO_PROTOCOL           *BlockIo;
+  EFI_DISK_IO_PROTOCOL            *DiskIo;
+  EFI_DEVICE_PATH                 *DevicePath;
+  UINT64                          FirstMftRecord;
+  NTFS_FILE                       *RootIndex;
+  NTFS_FILE                       *MftStart;
+} EFI_FS;
```

Comment:
I just realised some „UEFI structures“ (not on-disk) like this one are packed. Is this intentional and if so, why?

Reply:
It is so because all structures are put in one block for simplicity and historical reasons. I can move NTFS_ATTR, NTFS_FILE, EFI_NTFS_FILE, EFI_FS, UNIT_ELEMENT, COMPRESSED, RUNLIST out of pragma, if you insist.

Reply:
@vit9696 Do you need NTFS on any platform where this matters? If so, even some on-disk structures can be un-packed